### PR TITLE
Migrated to sqlalchemy-based db (DM-3511)

### PR DESCRIPTION
**Pull request by jbecla**

### CODE REVIEW

File: tests/metadataFitsTest.py [hunk 1]
```diff
@@ -84,11 +77,9 @@ def test_readInFits(self):
             rootDir = os.path.expanduser(rootDir)
         if not os.path.exists(rootDir):
             print "Data directory {} is required".format(rootDir)
```
Comment (timj) [inline tests/metadataFitsTest.py]: Tangential, but can we upgrade to `print_function` please? There are a few places in this package still using old style print.

File: python/lsst/dax/imgserv/MetadataFitsDb.py [hunk 2]
```diff
@@ -320,47 +308,40 @@ def insertMetadataFits(self, metadata):
...
-                    metadataPosition = MetadataPosition(lastFitsFileId, hdu, cursor, entries)
+                    metadataPosition = MetadataPosition(lastFitsFileId, hdu, self._conn, entries)
                     metadataPosition._insert()
-                cursor.execute("COMMIT")
-        except MySQLdb.Error as err:
-            cursor.execute("ROLLBACK")
+                trans.commit()
+        except SQLAlchemyError as e:
+            trans.rollback()
             self._log.warn( "ROLLBACK due to ERROR MySQLdb %s --%s", err, sql)
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/MetadataFitsDb.py]: `err` should become `e`

File: python/lsst/dax/imgserv/MetadataFitsDb.py [hunk 3]
```diff
@@ -320,47 +308,40 @@ def insertMetadataFits(self, metadata):
...
-                        self._insertFitsValue(cursor, lastFitsFileId, key, value, lineNum, comment)
+                        self._insertFitsValue(lastFitsFileId, key, value, lineNum, comment)
                 for hdu in range(1,hdus+1):
-                    metadataPosition = MetadataPosition(lastFitsFileId, hdu, cursor, entries)
+                    metadataPosition = MetadataPosition(lastFitsFileId, hdu, self._conn, entries)
                     metadataPosition._insert()
-                cursor.execute("COMMIT")
-        except MySQLdb.Error as err:
-            cursor.execute("ROLLBACK")
+                trans.commit()
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/MetadataFitsDb.py]: commit should happen outside `if` (if `len(r)` is >= 1 then you neither commit nor rollback, and I'm not sure what happens in this case).
I strongly recommend to use context management with transactions, this guarantees that transaction is properly closed (or aborted if exception happens). Check third example from http://docs.sqlalchemy.org/en/rel_1_0/core/connections.html#using-transactions

File: python/lsst/dax/imgserv/MetadataFitsDb.py [hunk 4]
```diff
@@ -405,15 +386,13 @@ def directoryCrawl(rootDir, metaDb):
...
     '''This test only works on specific servers and uses a large dataset.
     '''
-    credFile = "~/.lsst/dbAuth-dbServ.txt"
-    creds = readCredentialFile(credFile, log)
-    dbName = "{}_fitsTest".format(creds['user'])
+    credFile = "~/.lsst/dbAuth-dbServ.ini"
 
     # Destroy existing tables and re-create them
-    dbDestroyCreate(credFile, dbName, "DELETE")
+    dbName = dbDestroyCreate(credFile, "DELETE")
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/MetadataFitsDb.py]: pylint tells me that `dbDestroyCreate` always returns `None` so assigning it to `dbName` does not make sense, and `dbName` itself is not used anywhere, I think assignment is not needed.

File: python/lsst/dax/imgserv/MetadataFitsDb.py [hunk 5]
```diff
@@ -32,16 +32,15 @@
 import sys
 import time
 
+from sqlalchemy.exc import SQLAlchemyError
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/MetadataFitsDb.py]: There is `import MySQLdb` above which I think should disappear

File: python/lsst/dax/imgserv/MetadataFitsDb.py [hunk 6]
```diff
@@ -50,13 +49,15 @@ def isDateFormatValid(dt):
...
+def executeInsertList(conn, table, columnValues, logger=log):
     '''Insert the columnValues into 'table'
     columnValue is a list of column name and value pairs.
     Values are sanitized.
+
+    Returns lastrowid or None
     '''
     if len(columnValues) < 1:
-        return
+        return None
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/MetadataFitsDb.py]: Technically this is the same thing, but I agree that explicit `None` is better if function is expected to return something.

File: python/lsst/dax/imgserv/crawler.py [hunk 7]
```diff
@@ -83,10 +83,10 @@ def get_metadata(self, path):
 
 
     def run(self):
-        credFileName = "~/.lsst/dbAuth-dbServ.txt"
-        creds = readCredentialFile(credFileName, log)
-        dbName = "{}_fitsTest".format(creds['user'])
-        metaDb = dbOpen(credFileName, dbName)
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/crawler.py]: also remove `readCredentialFile` and `dbOpen` from imports

File: python/lsst/dax/imgserv/crawler.py [hunk 8]
```diff
@@ -83,10 +83,10 @@ def get_metadata(self, path):
...
+        credFileName = "~/.lsst/dbAuth-dbServ.ini"
+        engine = getEngineFromFile(credFileName)
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/crawler.py]: need import for `getEngineFromFile`

File: python/lsst/dax/imgserv/fitsMetadataSchema.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+(
+    fitsFileId BIGINT       NOT NULL AUTO_INCREMENT,
+    fileName   VARCHAR(255) NOT NULL,
+    hduCount   TINYINT      NOT NULL,
+    PRIMARY KEY(fitsFileId)
+) ENGINE=InnoDB
+""",
+
+"""CREATE TABLE FitsKeyValues
+    -- <descr>Table of FITS keyword and value pairs. </decr>
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/fitsMetadataSchema.py]: `</descr>` instead of `</decr>`

File: python/lsst/dax/imgserv/crawler.py [hunk 10]
```diff
@@ -83,10 +83,10 @@ def get_metadata(self, path):
...
+        dbName = "{}_fitsTest".format(engine.url.username)
+        metaDb = MetadataFitsDb(credFile)
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/crawler.py]: `credFileName` instead of `credFile`

File: python/lsst/dax/imgserv/fitsMetadataSchema.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    equinox    DOUBLE,
+    pDec       DOUBLE,
+    pRa        DOUBLE,
+    rotAng     DOUBLE,
+    pDate      TIMESTAMP,
+    INDEX IDX_fitsPos_fitsFileId(fitsFileId),
+    INDEX IDX_fitsPos_date(pDate),
+    INDEX IDX_fitsPos_ra(pRa),
+    INDEX IDX_fitsPos_dec(pDec)
+) ENGINE=InnoDB;
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/fitsMetadataSchema.py]: Semicolon not needed

File: python/lsst/dax/imgserv/locateImage.py [hunk 12]
```diff
@@ -55,27 +55,18 @@
...
-                 database, table, columns, dataRoot, logger):
+    def __init__(self, credFileName, database, table, columns, dataRoot, logger):
         self._log = logger
-        self._database = database
         self._table = table
         self._columns = columns
-        self._connect = MySQLdb.connect(host=dbHost, port=dbPort,
-                                        user=dbUser, passwd=dbPasswd,
-                                        db=self._database)
+        self._conn = getEngineFromFile(credFileName, database=database).connect()
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/locateImage.py]: Need an import for `getEngineFromFile`

File: python/lsst/dax/imgserv/locateImage.py [hunk 13]
```diff
@@ -44,9 +44,9 @@
 import lsst.afw.image as afwImage
 import lsst.daf.base as dafBase
 import lsst.log as log
+from sqlalchemy.exc import SQLAlchemyError
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/locateImage.py]: `import MySQLdb` above should be gone

File: python/lsst/dax/imgserv/locateImage.py [hunk 14]
```diff
@@ -250,7 +239,8 @@ class W13DeepCoaddDb(W13Db):
     '''
     def __init__(self, dbHost, dbPort, dbUser, dbPasswd, logger=log):
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/locateImage.py]: replace `__init__` arguments here too

File: tests/metadataFitsTest.py (hunk 1 above)
Comment (andy-slac) [inline tests/metadataFitsTest.py]: Idd say it should be replaced with `log.error` in this case

File: python/lsst/dax/imgserv/fitsMetadataSchema.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
+
+
```
Comment (jgates108) [inline python/lsst/dax/imgserv/fitsMetadataSchema.py]: No copyright info needed?

Comment (jbecla) [inline python/lsst/dax/imgserv/fitsMetadataSchema.py]: I just renamed this file and stuck the schema into a list. It wasn't there before :). I'll add it.

File: bin/imageServer.py [hunk 16]
```diff
@@ -21,15 +21,16 @@
...
+Use it for tests. It is really meant to run as part of the central
 Web Service, e.g., through webserv/bin/server.py
 
 @author  Jacek Becla, SLAC
 """
 
+import sys
 from flask import Flask
-from lsst.imgserv import imageREST_v0
+from lsst.dax.imgserv import imageREST_v0
```
Comment (brianv0) [inline bin/imageServer.py]: Was this not already fixed and on the master?

File: python/lsst/dax/imgserv/MetadataFitsDb.py [hunk 17]
```diff
@@ -294,73 +284,58 @@ def insertMetadataFits(self, metadata):
...
-                    metadataPosition._insert()
-                cursor.execute("COMMIT")
-        except MySQLdb.Error as err:
-            cursor.execute("ROLLBACK")
-            self._log.warn( "ROLLBACK due to ERROR MySQLdb %s --%s", err, sql)
-            quit() # TODO delete this line, for now it is good to stop and examine these
-        cursor.close()
+            with self._conn.begin() as trans:
+                sql = ("SELECT 1 FROM FitsFiles WHERE "
+                       "fileName = %s")
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/MetadataFitsDb.py]: Remove parens around string, should fit OK into one line

File: python/lsst/dax/imgserv/MetadataFitsDb.py [hunk 18]
```diff
@@ -294,73 +284,58 @@ def insertMetadataFits(self, metadata):
...
+                self._log.debug(sql, str(fileName))
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/MetadataFitsDb.py]: I'm not sure why `str()` is needed here

File: python/lsst/dax/imgserv/MetadataFitsDb.py [hunk 19]
```diff
@@ -294,73 +284,58 @@ def insertMetadataFits(self, metadata):
...
+                # Nothing found, so it needs to be added.
+                if len(r) < 1:
+                    #If this fails for any reason, we do not want the database altered.
+                    # Insert the file into the file table.
+                    colVal = [("fileName", fileName), ("hduCount", hdus)]
+                    lastFitsFileId = executeInsertList(self._conn, "FitsFiles", colVal, self._log)
+                    for key, entry in entries.iteritems():
+                        value   = entry[0]
+                        lineNum = entry[1]
+                        comment = entry[2]
```
Comment (andy-slac) [inline python/lsst/dax/imgserv/MetadataFitsDb.py]: I think more Pythonic is `value, lineNum, comment = entry` (or `entry[:3]` if entry can be longer)

File: python/lsst/dax/imgserv/fitsMetadataSchema.py (hunk 15 above)
Comment (andy-slac) [inline python/lsst/dax/imgserv/fitsMetadataSchema.py]: I'm really tired of the copyrights (and I'm definitely not reviewing them). I remember someone mentioning (at Bremerton?) that there should be enough to have one `COPYRIGHT` file at the top level, or was I dreaming?

Comment (timj) [inline python/lsst/dax/imgserv/fitsMetadataSchema.py]: You were not dreaming: https://jira.lsstcorp.org/browse/RFC-45

Comment (andy-slac) [inline python/lsst/dax/imgserv/fitsMetadataSchema.py]: Thanks @timj, it looks like we need a ticket to purge all copyrights from every file. I can review that :smile:

File: bin/imageServer.py (hunk 16 above)
Comment (jbecla) [inline bin/imageServer.py]: Notice that it is in dax_imgserv/bin/imageServer.py, it is not the same as dax_webserv/bin/server.py. We have "servers" for each module to test things locally. I think that one never got fixed.

File: python/lsst/dax/imgserv/MetadataFitsDb.py (hunk 18 above)
Comment (jbecla) [inline python/lsst/dax/imgserv/MetadataFitsDb.py]: I am not sure either, not my code :). I removed it

**Pull request merged**